src/Workspaces/Core/Portable/EmbeddedLanguages/Common/EmbeddedSyntaxTree.cs

```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.Common
+{
+    internal abstract class EmbeddedSyntaxTree<TSyntaxKind, TNode, TRoot>
+        where TSyntaxKind : struct
+        where TNode : EmbeddedSyntaxNode<TSyntaxKind, TNode>
+        where TRoot : TNode
```

Comment:
Is there a scenario where `TRoot` is not literally `TNode`?

Reply:
Yes.  TRoot is never literally TNode.  For regex and json it will be RegexCompilationUnit and JsonCompilationUnit respectively.

Reply:
Renamed to be clearer.  #resolved.